trace/collector.go

```diff
@@ -0,0 +1,49 @@
+package trace
+
+import (
+	"log"
+	"sync"
+)
+
+type Collector interface {
+	Record(e Event) error
+}
+
+type MemoryCollector struct {
+	list []Event
+	in   chan Event
+	lock sync.Mutex
+}
+
+func (c MemoryCollector) Record(event Event) error {
+	c.in <- event
+	return nil
+}
+
+func (c MemoryCollector) GetEvents() []Event {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	listLength := len(c.list)
+	newList := make([]Event, listLength)
+	copied := copy(newList, c.list)
+	if copied != listLength {
+		log.Panicln("Failed to copy Events from list")
+	}
+	return newList
+}
+
+func NewMemoryCollector() *MemoryCollector {
```

Comment:
Only a nitpicking but I think that it is better to put the constructor function directly below th type definition:

```go

type A struct {...}

func NewA(...) *A { ... }
```

I think it's easier to grok the given type.